File: packages/claude-agent-sdk-adapter/src/adapter.ts
import type {
  TextConfig,
  ImageConfig,
  ObjectConfig,
  SpeechConfig,
  PromptMetadata,
  AdaptOptions,
  PromptShape,
  KeysWithKind,
  Adapter,
  RichChatMessage,
} from "@agentmark-ai/prompt-core";
import { ClaudeAgentModelRegistry } from "./model-registry";
import { ClaudeAgentToolRegistry } from "./tool-registry";
import {
  createAgentMarkMcpServer,
  toClaudeAgentMcpServer,
} from "./mcp/agentmark-mcp-bridge";
import type {
  ClaudeAgentTextParams,
  ClaudeAgentObjectParams,
  ClaudeAgentAdapterOptions,
  ClaudeAgentQueryOptions,
  McpServerConfig,
  TracedTelemetryContext,
} from "./types";

/**
 * Config options supported by Claude Agent SDK adapter for text prompts.
 * Other options will trigger a warning when present.
 */
const SUPPORTED_TEXT_OPTIONS = new Set([
  "model_name", // Used via model registry
  "max_calls",  // Mapped to maxTurns
  "tools",      // Converted to MCP servers
]);

/**
 * Config options supported by Claude Agent SDK adapter for object prompts.
 * Includes all text options plus schema-related options.
 */
const SUPPORTED_OBJECT_OPTIONS = new Set([
  ...SUPPORTED_TEXT_OPTIONS,
  "schema",             // Used for outputFormat
  "schema_name",        // Passed through for schema naming (optional)
  "schema_description", // Passed through for schema description (optional)
]);

/**
 * Check for unsupported config options and emit warnings.
 *
 * @param settings - The settings object from the prompt config
 * @param supportedOptions - Set of supported option names
 * @param promptName - Name of the prompt for the warning message
 * @param configType - Type of config (text_config or object_config) for the warning message
 */
function warnUnsupportedOptions(
  settings: Record<string, unknown>,
  supportedOptions: Set<string>,
  promptName: string,
  configType: "text_config" | "object_config"
): void {
  const unsupportedOptions = Object.keys(settings).filter(
    (key) => !supportedOptions.has(key)
  );

  if (unsupportedOptions.length > 0) {
    console.warn(
      `[claude-agent-sdk-adapter] Warning: The following ${configType} options in prompt "${promptName}" are not supported by Claude Agent SDK and will be ignored: ${unsupportedOptions.join(", ")}`
    );
  }
}

/**
 * AgentMark adapter for Claude Agent SDK.
 *
 * This adapter bridges AgentMark's prompt framework with Anthropic's Claude Agent SDK,
 * enabling autonomous agent execution with built-in tools.
 *
 * Key differences from other adapters:
 * - Execution is agentic (autonomous loop) rather than request-response
 * - Uses AsyncGenerator streaming instead of ReadableStream
 * - Has built-in tools (Read, Write, Bash, etc.)
 * - Includes permission system for tool access control
 *
 * @example
 * ```typescript
 * import { ClaudeAgentAdapter, ClaudeAgentModelRegistry } from "@agentmark-ai/claude-agent-sdk-adapter";
 *
 * const adapter = new ClaudeAgentAdapter(
 *   ClaudeAgentModelRegistry.createDefault(),
 *   toolRegistry,
 *   { permissionMode: 'bypassPermissions' }
 * );
 * ```
 */
export class ClaudeAgentAdapter<
  T extends PromptShape<T>,
  R extends ClaudeAgentToolRegistry<any, any> = ClaudeAgentToolRegistry<any, any>
> implements Adapter<T>
{
  declare readonly __dict: T;
  readonly __name = "claude-agent-sdk";

  constructor(
    private modelRegistry: ClaudeAgentModelRegistry,
    private toolRegistry?: R,
    private adapterOptions?: ClaudeAgentAdapterOptions
  ) {}

  /**
   * Convert RichChatMessages to a prompt string for Claude Agent SDK.
   *
   * Claude Agent SDK expects a prompt string, not message arrays.
   * System messages are handled separately via the systemPrompt option.
   */
  private messagesToPrompt(messages: RichChatMessage[]): string {
    return messages
      .filter((m) => m.role !== "system")
      .map((m) => {
        if (typeof m.content === "string") {
          return m.content;
        }
        // Handle rich content (array of parts)
        return m.content
          .map((part) => {
            if (part.type === "text") {
              return part.text;
            }
            // For file/image parts, include a placeholder
            if (part.type === "file" || part.type === "image") {
              return `[Attached ${part.type}]`;
            }
            return "";
          })
          .filter(Boolean)
          .join("\n");
      })
      .join("\n\n");
  }

  /**
   * Extract system prompt from messages.
   */
  private extractSystemPrompt(messages: RichChatMessage[]): string | undefined {
    const systemMsg = messages.find((m) => m.role === "system");
    if (!systemMsg) return undefined;

    if (typeof systemMsg.content === "string") {
      return systemMsg.content;
    }

    // Handle rich content
    const content = systemMsg.content as Array<{ type: string; text?: string }>;
    return content
      .filter((part: { type: string }) => part.type === "text")
      .map((part: { type: string; text?: string }) => part.text ?? "")
      .join("\n");
  }

  /**
   * Convert tools defined in prompt frontmatter to AgentMark tool definitions.
   * Only includes tools that have a registered executor in the tool registry.
   */
  private convertPromptToolsToAgentMarkTools(
    tools: Record<string, { description: string; parameters: Record<string, unknown> }>
  ): Array<{ name: string; description: string; parameters: Record<string, unknown>; execute: (args: unknown) => Promise<unknown> }> {
    return Object.entries(tools)
      .filter(([name]) => this.toolRegistry?.has(name))
      .map(([name, toolDef]) => {
        const executor = this.toolRegistry!.get(name as never) as unknown as (args: unknown) => unknown;
        return {
          name,
          description: toolDef.description,
          parameters: toolDef.parameters,
          execute: async (args: unknown) => executor(args),
        };
      });
  }

  /**
   * Build common query options from config and adapter settings.
   * Returns both the query options and telemetry context (if telemetry enabled).
   */
  private buildQueryOptions(
    modelConfig: { model: string; maxThinkingTokens?: number },
    systemPrompt: string | undefined,
    settings: Record<string, unknown>,
    options: AdaptOptions,
    metadata: PromptMetadata,
    promptName: string
  ): { queryOptions: ClaudeAgentQueryOptions; telemetry?: TracedTelemetryContext } {
    // Build MCP servers for AgentMark tools
    // Use McpServerConfig type for compatibility with Claude Agent SDK
    const mcpServers: Record<string, McpServerConfig> = {};

    // Add tools defined in the prompt's frontmatter
    // Tool executors are matched from the tool registry
    if (settings.tools && typeof settings.tools === 'object') {
      const promptTools = this.convertPromptToolsToAgentMarkTools(
        settings.tools as Record<string, { description: string; parameters: Record<string, unknown> }>
      );
      if (promptTools.length > 0) {
        const mcpServer = createAgentMarkMcpServer("prompt-tools", promptTools);
        mcpServers["prompt-tools"] = toClaudeAgentMcpServer(mcpServer) as unknown as McpServerConfig;
      }
    }

    // Build telemetry context for withTracing() wrapper (only if telemetry enabled)
    let telemetry: TracedTelemetryContext | undefined;

    if (options.telemetry?.isEnabled) {
      telemetry = {
        isEnabled: true,
        promptName,
        systemPrompt,
        model: modelConfig.model,
        props: metadata.props,
        metadata: options.telemetry.metadata,
      };
    }

    // Build system prompt configuration
    let systemPromptConfig: string | { type: 'preset'; preset: 'claude_code'; append?: string } | undefined;
    if (this.adapterOptions?.systemPromptPreset) {
      systemPromptConfig = {
        type: "preset" as const,
        preset: "claude_code" as const,
        ...(systemPrompt && { append: systemPrompt }),
      };
    } else if (systemPrompt) {
      systemPromptConfig = systemPrompt;
    }

    const queryOptions: ClaudeAgentQueryOptions = {
      model: modelConfig.model,
      ...(modelConfig.maxThinkingTokens && {
        maxThinkingTokens: modelConfig.maxThinkingTokens,
      }),
      ...((settings as Record<string, unknown>).max_calls ? { maxTurns: (settings as Record<string, unknown>).max_calls as number } : {}),
      ...(this.adapterOptions?.permissionMode && {
        permissionMode: this.adapterOptions.permissionMode,
      }),
      ...(this.adapterOptions?.cwd && { cwd: this.adapterOptions.cwd }),
      ...(this.adapterOptions?.maxBudgetUsd && {
        maxBudgetUsd: this.adapterOptions.maxBudgetUsd,
      }),
      ...(this.adapterOptions?.maxTurns && {
        maxTurns: this.adapterOptions.maxTurns,
      }),
      ...(Object.keys(mcpServers).length > 0 && { mcpServers }),
      ...(systemPromptConfig && { systemPrompt: systemPromptConfig }),
      ...(this.adapterOptions?.allowedTools && {
        allowedTools: this.adapterOptions.allowedTools,
      }),
      ...(this.adapterOptions?.disallowedTools && {
        disallowedTools: this.adapterOptions.disallowedTools,
      }),
    };

    return { queryOptions, telemetry };
  }

  /**
   * Adapt a text configuration for Claude Agent SDK.
   *
   * @param input - Text configuration from AgentMark prompt
   * @param options - Adapt options including telemetry settings
   * @param metadata - Prompt metadata including props
   * @returns Configuration for Claude Agent SDK query() with telemetry context
   */
  async adaptText<_K extends KeysWithKind<T, "text"> & string>(
    input: TextConfig,
    options: AdaptOptions,
    metadata: PromptMetadata
  ): Promise<ClaudeAgentTextParams> {
    const { model_name, ...settings } = input.text_config;

    // Warn about unsupported config options
    warnUnsupportedOptions(settings, SUPPORTED_TEXT_OPTIONS, input.name, "text_config");

    const modelConfig = this.modelRegistry.getModelConfig(model_name, options);
    const systemPrompt = this.extractSystemPrompt(input.messages);
    const userPrompt = this.messagesToPrompt(input.messages);

    const { queryOptions, telemetry } = this.buildQueryOptions(
      modelConfig,
      systemPrompt,
      settings,
      options,
      metadata,
      input.name
    );

    return {
      query: {
        prompt: userPrompt,
        options: queryOptions,
      },
      messages: input.messages,
      telemetry,
    };
  }

  /**
   * Adapt an object configuration for Claude Agent SDK with structured output.
   *
   * @param input - Object configuration from AgentMark prompt
   * @param options - Adapt options including telemetry settings
   * @param metadata - Prompt metadata including props
   * @returns Configuration for Claude Agent SDK query() with outputFormat and telemetry context
   */
  async adaptObject<K extends KeysWithKind<T, "object"> & string>(
    input: ObjectConfig,
    options: AdaptOptions,
    metadata: PromptMetadata
  ): Promise<ClaudeAgentObjectParams<T[K]["output"]>> {
    const { model_name, schema, ...settings } = input.object_config;

    // Warn about unsupported config options
    warnUnsupportedOptions(settings, SUPPORTED_OBJECT_OPTIONS, input.name, "object_config");

    const modelConfig = this.modelRegistry.getModelConfig(model_name, options);
    const systemPrompt = this.extractSystemPrompt(input.messages);
    const userPrompt = this.messagesToPrompt(input.messages);

    const { queryOptions, telemetry } = this.buildQueryOptions(
      modelConfig,
      systemPrompt,
      settings,
      options,
      metadata,
      input.name
    );

    // Add structured output format
    const optionsWithOutput = {
      ...queryOptions,
      outputFormat: {
        type: "json_schema" as const,
        schema: schema,
      },
    };

    return {
      query: {
        prompt: userPrompt,
        options: optionsWithOutput,
      },
      messages: input.messages,
      telemetry,
    };
  }

  /**
   * Image generation is not supported by Claude Agent SDK.
   *
   * @throws Error with guidance to use a different adapter
   */
  adaptImage<_K extends KeysWithKind<T, "image"> & string>(
    _input: ImageConfig,
    _options: AdaptOptions
  ): never {
    throw new Error(
      "Image generation is not supported by Claude Agent SDK. " +
        "Consider using the Vercel AI SDK adapter with an image model like DALL-E or Stable Diffusion."
    );
  }

  /**
   * Speech generation is not supported by Claude Agent SDK.
   *
   * @throws Error with guidance to use a different adapter
   */
  adaptSpeech<_K extends KeysWithKind<T, "speech"> & string>(
    _input: SpeechConfig,
    _options: AdaptOptions
  ): never {
    throw new Error(
      "Speech generation is not supported by Claude Agent SDK. " +
        "Consider using the Vercel AI SDK adapter with a speech model like OpenAI TTS."
    );
  }
}

File: packages/claude-agent-sdk-adapter-python/src/agentmark_claude_agent_sdk/adapter.py
"""AgentMark adapter for Claude Agent SDK.

Ported from TypeScript: packages/claude-agent-sdk-adapter/src/adapter.ts
"""

from __future__ import annotations

from typing import Any

from .mcp.agentmark_mcp_bridge import (
    AgentMarkToolDefinition,
    create_agentmark_mcp_server,
    to_claude_agent_mcp_server,
)
from .model_registry import ClaudeAgentModelRegistry
from .tool_registry import ClaudeAgentToolRegistry
from .types import (
    ClaudeAgentAdapterOptions,
    ClaudeAgentObjectParams,
    ClaudeAgentQueryOptions,
    ClaudeAgentQueryParams,
    ClaudeAgentTextParams,
    OutputFormat,
    SystemPromptPreset,
    TracedTelemetryContext,
)


class ClaudeAgentAdapter:
    """AgentMark adapter for Claude Agent SDK.

    This adapter bridges AgentMark's prompt framework with Anthropic's Claude Agent SDK,
    enabling autonomous agent execution with built-in tools.

    Key differences from other adapters:
    - Execution is agentic (autonomous loop) rather than request-response
    - Uses AsyncGenerator streaming instead of ReadableStream
    - Has built-in tools (Read, Write, Bash, etc.)
    - Includes permission system for tool access control

    Example:
        from agentmark_claude_agent_sdk import (
            ClaudeAgentAdapter,
            create_default_model_registry,
        )

        adapter = ClaudeAgentAdapter(
            model_registry=create_default_model_registry(),
            tool_registry=tool_registry,
            adapter_options=ClaudeAgentAdapterOptions(permission_mode='bypassPermissions'),
        )
    """

    name = "claude-agent-sdk"

    def __init__(
        self,
        model_registry: ClaudeAgentModelRegistry,
        tool_registry: ClaudeAgentToolRegistry | None = None,
        adapter_options: ClaudeAgentAdapterOptions | None = None,
    ) -> None:
        """Initialize the adapter.

        Args:
            model_registry: Registry for model configurations.
            tool_registry: Optional registry for tool executors.
            adapter_options: Optional adapter-level options.
        """
        self._model_registry = model_registry
        self._tool_registry = tool_registry
        self._adapter_options = adapter_options

    def _messages_to_prompt(self, messages: list[dict[str, Any]]) -> str:
        """Convert RichChatMessages to a prompt string for Claude Agent SDK.

        Claude Agent SDK expects a prompt string, not message arrays.
        System messages are handled separately via the systemPrompt option.
        """
        result = []
        for m in messages:
            if m.get("role") == "system":
                continue

            content = m.get("content")
            if isinstance(content, str):
                result.append(content)
            elif isinstance(content, list):
                # Handle rich content (array of parts)
                parts = []
                for part in content:
                    if part.get("type") == "text":
                        parts.append(part.get("text", ""))
                    elif part.get("type") in ("file", "image"):
                        parts.append(f"[Attached {part.get('type')}]")
                result.append("\n".join(filter(None, parts)))

        return "\n\n".join(result)

    def _extract_system_prompt(self, messages: list[dict[str, Any]]) -> str | None:
        """Extract system prompt from messages."""
        for m in messages:
            if m.get("role") == "system":
                content = m.get("content")
                if isinstance(content, str):
                    return content
                elif isinstance(content, list):
                    # Handle rich content
                    texts = []
                    for part in content:
                        if part.get("type") == "text":
                            texts.append(part.get("text", ""))
                    return "\n".join(texts) if texts else None
        return None

    def _convert_prompt_tools_to_agentmark_tools(
        self, tools: dict[str, dict[str, Any]]
    ) -> list[AgentMarkToolDefinition]:
        """Convert tools defined in prompt frontmatter to AgentMark tool definitions.

        Only includes tools that have a registered executor in the tool registry.
        """
        if not self._tool_registry:
            return []

        result = []
        for name, tool_def in tools.items():
            if not self._tool_registry.has(name):
                continue

            executor = self._tool_registry.get(name)
            if executor is None:
                continue

            async def execute_wrapper(args: dict[str, Any], _executor: Any = executor) -> Any:
                return await _executor.execute(args, None)

            result.append(
                AgentMarkToolDefinition(
                    name=name,
                    description=tool_def.get("description", ""),
                    parameters=tool_def.get("parameters", {}),
                    execute=execute_wrapper,
                )
            )

        return result

    def _build_query_options(
        self,
        model_config: dict[str, Any],
        system_prompt: str | None,
        settings: dict[str, Any],
        options: dict[str, Any],
        metadata: dict[str, Any],
        prompt_name: str,
    ) -> tuple[ClaudeAgentQueryOptions, TracedTelemetryContext | None]:
        """Build common query options from config and adapter settings.

        Returns both the query options and telemetry context (if telemetry enabled).
        """
        # Build MCP servers for AgentMark tools
        mcp_servers: dict[str, Any] = {}

        # Add tools defined in the prompt's frontmatter
        tools = settings.get("tools")
        if tools and isinstance(tools, dict):
            prompt_tools = self._convert_prompt_tools_to_agentmark_tools(tools)
            if prompt_tools:
                mcp_server = create_agentmark_mcp_server("prompt-tools", prompt_tools)
                mcp_servers["prompt-tools"] = to_claude_agent_mcp_server(mcp_server)

        # Build telemetry context for withTracing() wrapper (only if telemetry enabled)
        telemetry: TracedTelemetryContext | None = None

        telemetry_options = options.get("telemetry", {})
        if telemetry_options.get("isEnabled"):
            telemetry = TracedTelemetryContext(
                is_enabled=True,
                prompt_name=prompt_name,
                system_prompt=system_prompt,
                model=model_config.get("model"),
                props=metadata.get("props"),
                metadata=telemetry_options.get("metadata"),
            )

        # Build system prompt configuration
        system_prompt_config: str | SystemPromptPreset | None = None
        if self._adapter_options and self._adapter_options.system_prompt_preset:
            system_prompt_config = SystemPromptPreset(
                type="preset",
                preset="claude_code",
                append=system_prompt if system_prompt else None,
            )
        elif system_prompt:
            system_prompt_config = system_prompt

        # Build query options
        query_options = ClaudeAgentQueryOptions(
            model=model_config.get("model"),
            max_thinking_tokens=model_config.get("max_thinking_tokens"),
            max_turns=settings.get("max_calls")
            or (self._adapter_options.max_turns if self._adapter_options else None),
            permission_mode=self._adapter_options.permission_mode
            if self._adapter_options
            else None,
            cwd=self._adapter_options.cwd if self._adapter_options else None,
            max_budget_usd=self._adapter_options.max_budget_usd if self._adapter_options else None,
            mcp_servers=mcp_servers if mcp_servers else None,
            system_prompt=system_prompt_config,
            allowed_tools=self._adapter_options.allowed_tools if self._adapter_options else None,
            disallowed_tools=self._adapter_options.disallowed_tools
            if self._adapter_options
            else None,
        )

        return query_options, telemetry

    async def adapt_text(
        self,
        config: dict[str, Any],
        options: dict[str, Any],
        messages: list[dict[str, Any]],
        metadata: dict[str, Any],
    ) -> ClaudeAgentTextParams:
        """Adapt a text configuration for Claude Agent SDK.

        Args:
            config: Text configuration from AgentMark prompt.
            options: Adapt options including telemetry settings.
            messages: Chat messages.
            metadata: Prompt metadata including props.

        Returns:
            Configuration for Claude Agent SDK query() with telemetry context.
        """
        text_config = config.get("text_config", {})
        model_name = text_config.get("model_name", "")
        settings = {k: v for k, v in text_config.items() if k != "model_name"}

        model_config_obj = self._model_registry.get_model_config(model_name, options)
        model_config = {
            "model": model_config_obj.model,
            "max_thinking_tokens": model_config_obj.max_thinking_tokens,
        }

        system_prompt = self._extract_system_prompt(messages)
        user_prompt = self._messages_to_prompt(messages)
        prompt_name = config.get("name", metadata.get("name", ""))

        query_options, telemetry = self._build_query_options(
            model_config, system_prompt, settings, options, metadata, prompt_name
        )

        return ClaudeAgentTextParams(
            query=ClaudeAgentQueryParams(prompt=user_prompt, options=query_options),
            messages=messages,
            telemetry=telemetry,
            prompt_name=prompt_name,
        )

    async def adapt_object(
        self,
        config: dict[str, Any],
        options: dict[str, Any],
        messages: list[dict[str, Any]],
        metadata: dict[str, Any],
    ) -> ClaudeAgentObjectParams[Any]:
        """Adapt an object configuration for Claude Agent SDK with structured output.

        Args:
            config: Object configuration from AgentMark prompt.
            options: Adapt options including telemetry settings.
            messages: Chat messages.
            metadata: Prompt metadata including props.

        Returns:
            Configuration for Claude Agent SDK query() with outputFormat and telemetry context.
        """
        object_config = config.get("object_config", {})
        model_name = object_config.get("model_name", "")
        schema = object_config.get("schema", {})
        settings = {k: v for k, v in object_config.items() if k not in ("model_name", "schema")}

        model_config_obj = self._model_registry.get_model_config(model_name, options)
        model_config = {
            "model": model_config_obj.model,
            "max_thinking_tokens": model_config_obj.max_thinking_tokens,
        }

        system_prompt = self._extract_system_prompt(messages)
        user_prompt = self._messages_to_prompt(messages)
        prompt_name = config.get("name", metadata.get("name", ""))

        query_options, telemetry = self._build_query_options(
            model_config, system_prompt, settings, options, metadata, prompt_name
        )

        # Add structured output format
        query_options.output_format = OutputFormat(type="json_schema", schema=schema)

        return ClaudeAgentObjectParams(
            query=ClaudeAgentQueryParams(prompt=user_prompt, options=query_options),
            messages=messages,
            output_schema=schema,
            telemetry=telemetry,
            prompt_name=prompt_name,
        )

    def adapt_image(self, config: dict[str, Any], options: dict[str, Any]) -> None:
        """Image generation is not supported by Claude Agent SDK.

        Raises:
            NotImplementedError: Always raises with guidance to use a different adapter.
        """
        raise NotImplementedError(
            "Image generation is not supported by Claude Agent SDK. "
            "Consider using a different adapter with an image model like DALL-E or Stable Diffusion."
        )

    def adapt_speech(self, config: dict[str, Any], options: dict[str, Any]) -> None:
        """Speech generation is not supported by Claude Agent SDK.

        Raises:
            NotImplementedError: Always raises with guidance to use a different adapter.
        """
        raise NotImplementedError(
            "Speech generation is not supported by Claude Agent SDK. "
            "Consider using a different adapter with a speech model like OpenAI TTS."
        )
